server/rw_wrapper.go

```diff
@@ -0,0 +1,63 @@
+package server
+
+import (
+	"bytes"
+	"compress/gzip"
+	"net/http"
+
+	"github.com/avenga/couper/handler"
+)
+
+var _ http.ResponseWriter = &RWWrapper{}
+
+// RWWrapper wraps the <http.ResponseWriter>.
+type RWWrapper struct {
+	rw    http.ResponseWriter
+	useGZ bool
+}
+
+// NewRWWrapper creates a new RWWrapper object.
+func NewRWWrapper(rw http.ResponseWriter, useGZ bool) *RWWrapper {
+	return &RWWrapper{
+		rw:    rw,
+		useGZ: useGZ,
+	}
+}
+
+// Header wraps the Header method of the <http.ResponseWriter>.
+func (w *RWWrapper) Header() http.Header {
+	return w.rw.Header()
+}
+
+// Write wraps the Write method of the <http.ResponseWriter>.
+func (w *RWWrapper) Write(p []byte) (int, error) {
+	if w.useGZ {
+		var buf bytes.Buffer
+
+		gz := gzip.NewWriter(&buf)
+		defer gz.Close()
+
+		if n, err := gz.Write(p); err != nil {
+			return n, err
+		}
+
+		gz.Close()
```

Comment:
Additionally this would not work at all. `Write` could be called multiple times during a `io.Copy()` for example. Gzip initialisation and close/flush must be done outside of `Write`.

See https://golang.org/pkg/net/http/#Flusher to trigger this within a test origin and in general with `gz.Close` (Flush) and the underlying http rw (flusher).